File: DIAL_Practica1_Parte1.py


'''

    PRÁCTICA 1: ANÁLISIS DE LA EFICIENCIA DE ALGORITMOS (PYTHON)

    ASIGNATURA: DISEÑO DE ALGORITMOS
    CURSO:      2022-2023
    SUBGRUPO:   U1

    APELLIDOS: Alba Serrano
    NOMBRE: Gabriel
        
'''



import matplotlib.pyplot as plt
import numpy as np
import timeit
import time
import random



# EJEMPLO: MÁXIMO COMÚN DIVISOR



def mcd_rec(a, b):
    
    if a > b :
        return mcd_rec(a - b, b)
    elif a == b :
        return a
    else :
        return mcd_rec(a, b - a)



def mcd_rec2(a, b):
    
    if b == 0 :
        return a
    else :
        return mcd_rec2(b, a % b)
    
    

def mcd_iter(a, b):
    
    a1 = a
    b1 = b
    
    while a1 != b1 :
        
        if a1 > b1 :
            a1 = a1 - b1
        else :
            b1 = b1 - a1
            
    return a1



def main():
    
    MAX_LEN = 500  # Maximum length of input list.

    # Initialise results containers:
    
    lengths_mcd_rec  = []
    times_mcd_rec    = []
    
    lengths_mcd_rec2 = []
    times_mcd_rec2   = []
    
    lengths_mcd_iter = []
    times_mcd_iter   = []
    

    for length in range(0, MAX_LEN, 10):
        
        # Generate random values:
        
        a = random.randint(length, 10*length)
        b = random.randint(length, 10*length)

        # Time execution (mcd_rec):
        
        start = time.perf_counter()
        mcd_rec(a, b)
        end = time.perf_counter()

        # Store results (mcd_rec):
        
        lengths_mcd_rec.append(length)
        times_mcd_rec.append(end - start)

        # Time execution (mcd_rec2):
        
        start = time.perf_counter()
        mcd_rec2(a, b)
        end = time.perf_counter()

        # Store results (mcd_rec2):
        
        lengths_mcd_rec2.append(length)
        times_mcd_rec2.append(end - start)
        
        # Time execution (mcd_iter):
        
        start = time.perf_counter()
        mcd_iter(a, b)
        end = time.perf_counter()

        # Store results (mcd_iter):
        
        lengths_mcd_iter.append(length)
        times_mcd_iter.append(end - start)
        
        
        
    # Plot results
    
    plt.style.use("dark_background")
    plt.figure().canvas.manager.set_window_title("Algoritmos de MCD - Time Complexity")
    plt.xlabel("List Length")
    plt.ylabel("Execution Time (s)")
    plt.plot(lengths_mcd_rec, times_mcd_rec, label="mcd_rec()")
    plt.plot(lengths_mcd_rec2, times_mcd_rec2, label="mcd_rec2()")
    plt.plot(lengths_mcd_iter, times_mcd_iter, label="mcd_iter()")
    plt.legend()
    plt.tight_layout()
    plt.show()
    
    
    # Polynomial fit
    
    ns = np.linspace(1, 100000, 100, dtype = int)
    ts = [timeit.timeit('mcd_rec2(lst[0], lst[len(lst) - 1])',
                    setup='lst=list(range({})); random.shuffle(lst)'.format(n),
                    globals=globals(),
                    number=1000)
          for n in ns]

    plt.plot(ns, ts, 'or')

    degree = 5
    coeffs = np.polyfit(ns, ts, degree)
    p = np.poly1d(coeffs)
    plt.plot(ns, [p(n) for n in ns], '-b')



'''

EJERCICIO: Programa en Python los algoritmos 'queHace1', 'queHace2' y 'queHace3'
que has programado anteriormente en C++. Usa el código anterior para visualizar 
los tiempos de ejecución de estos tres programas, para determinar cuál es más 
eficiente en la práctica, así como para visualizar la función de coste que 
define, para cada algoritmo, su orden de complejidad.

'''

def queHace1(v,N): # O(N**3)
    mejor = 0 # O(1)
    for i in range(0,N): # O(N**3)
        for j in range(i,N): # O(suma : i <= j <= N-1 : j-i+1) = O(N**2)
            suma = 0 # O(1)
            for k in range(i,j+1): # O(j-i+1)
                suma += v[k] # O(1)
            if suma > mejor: # O(1)
                mejor = suma
    return mejor

def queHace2(v,N): # O(N**2)
    mejor = 0 # O(1)
    for i in range(0,N): # O(suma : 0 <= i <= N-1 : N-i) = O(N**2)
        suma = 0 # O(1)
        for j in range(i,N): # O(N-i)
            suma += v[j] # O(1)
            if suma > mejor: # O(1)
                mejor = suma # O(1)
    return mejor

def queHace3(v,N): # O(N)
    r, s = v[0], v[0] # O(1)
    for i in range(1,N): # O(N-1-1) = O(N-2)
        s = max(s+v[i] , v[i]) # O(1)
        r = max(r,s) # O(1)
    return r
    


MAX_LEN = 50  # Maximum length of input list.

# Initialise results containers:

lengths_queHace1  = []
times_queHace1    = []

lengths_queHace2 = []
times_queHace2   = []

lengths_queHace3 = []
times_queHace3   = []


for length in range(1, MAX_LEN):
    
    # Generate random values:
    
    v = random.sample(range(-100,100),length)

    # Time execution (queHace1):
    
    start = time.perf_counter()
    queHace1(v,length)
    end = time.perf_counter()

    # Store results (queHace1):
    
    lengths_queHace1.append(length)
    times_queHace1.append(end - start)

    # Time execution (mcd_rec2):
    
    start = time.perf_counter()
    queHace2(v,length)
    end = time.perf_counter()

    # Store results (mcd_rec2):
    
    lengths_queHace2.append(length)
    times_queHace2.append(end - start)
    
    # Time execution (mcd_iter):
    
    start = time.perf_counter()
    queHace3(v,length)
    end = time.perf_counter()

    # Store results (mcd_iter):
    
    lengths_queHace3.append(length)
    times_queHace3.append(end - start)
    
    
    
# Plot results

plt.style.use("dark_background")
plt.figure().canvas.manager.set_window_title("Algoritmos 'queHace' - Time Complexity")
plt.xlabel("List Length")
plt.ylabel("Execution Time (s)")
plt.plot(lengths_queHace1, times_queHace1, label="queHace1")
plt.plot(lengths_queHace2, times_queHace2, label="queHace2")
plt.plot(lengths_queHace3, times_queHace3, label="queHace3")
plt.legend()
plt.tight_layout()
plt.show()


# Polynomial fit for 'queHace1'

ns = np.linspace(1, 100000, 20, dtype = int)
ts = [timeit.timeit('queHace1(lst, len(lst) )',
                setup='lst=random.sample(range(-100,100),{})'.format(n),
                globals=globals(),
                number=1000)
      for n in range(1,len(ns)+1)]

plt.plot(ns, ts, 'or')

degree = 5
coeffs = np.polyfit(ns, ts, degree)
p = np.poly1d(coeffs)
plt.plot(ns, [p(n) for n in ns], label="queHace1")


# Polynomial fit for 'queHace2'

ns = np.linspace(1, 100000, 20, dtype = int)
ts = [timeit.timeit('queHace2(lst, len(lst) )',
                setup='lst=random.sample(range(-100,100),{})'.format(n),
                globals=globals(),
                number=1000)
      for n in range(1,len(ns)+1)]

plt.plot(ns, ts, 'or')

degree = 5
coeffs = np.polyfit(ns, ts, degree)
p = np.poly1d(coeffs)
plt.plot(ns, [p(n) for n in ns], label="queHace2")


# Polynomial fit for 'queHace3'

ns = np.linspace(1, 100000, 20, dtype = int)
ts = [timeit.timeit('queHace3(lst, len(lst) )',
                setup='lst=random.sample(range(-100,100),{})'.format(n),
                globals=globals(),
                number=1000)
      for n in range(1,len(ns)+1)]

plt.plot(ns, ts, 'or')

degree = 5
coeffs = np.polyfit(ns, ts, degree)
p = np.poly1d(coeffs)
plt.plot(ns, [p(n) for n in ns], label="queHace3")
plt.legend()

File: DIAL_Practica1_Parte1.cpp
#include <iostream>
#include <stdio.h>
#include <time.h>


using namespace std ;


/*

    PRÁCTICA 1: ANÁLISIS DE LA EFICIENCIA DE LOS ALGORITMOS (C/C++)

    ASIGNATURA: DISEÑO DE ALGORITMOS
    CURSO:      2022-2023
    SUBGRUPO:   U1

    APELLIDOS: Alba Serrano
    NOMBRE: Gabriel

*/



int queHace1 (int v[], int N){
    
  int mejor, suma ;
  
  mejor = 0 ;
  
  for (int i = 0 ; i < N ; i++){
      
    for (int j = i ; j < N ; j++){
        
       suma = 0 ;
       
       for (int k = i ; k <= j ; k++) suma = suma + v[k] ;
       
       if (suma > mejor) mejor = suma ;
       
    }
    
  }
  
  return mejor ;
  
}



/*
 
 EJERCICIO 1: Escribe el código en C++ de un algoritmo 'queHace2'
 que resuelva el mismo problema que el del algoritmo anterior
 'queHace1' pero usando solo dos bucles anidados.
 
 */

int queHace2 (int v[], int N){ // O(N^2)
    
  int mejor, suma ; // O(1)
  
  mejor = 0 ; // O(1)
  
  for (int i = 0 ; i < N ; i++){ // O( Sum_{i=0}^{N-1} (N-i) ) = O( N*(N+1)/2 ) = O(N^2)
    
    suma = 0 ; // O(1)
    
    for (int j = i ; j < N ; j++){ // O(N - i)
        
       suma = suma + v[j] ; // O(1)
       
       if (suma > mejor) mejor = suma ; // O(1)
       
    }
    
  }
  
  return mejor ; // O(1)
  
}

/*
 
 EJERCICIO 2: Escribe el código en C++ de un algoritmo 'queHace3'
 que resuelva el mismo problema que el del algoritmo anterior
 'queHace1' pero usando solo un bucle.
 
 */

int queHace3 (int v[], int N){ // O(N-1)
    
  int mejor, suma ; // O(1)
  suma = v[0] ; // O(1)
  mejor = v[0] ; // O(1)
  
  for (int i = 1 ; i < N ; i++){ // O(N-1)
  	
    suma = max(suma+v[i],v[i]) ; // O(1)
    mejor = max(mejor,suma) ; // O(1)
    
  }
  
  return mejor ; // O(1)
  
}

// Otra forma de definir la funcion 'queHace3' seria:

int queHace3b(int v[], int N) { // O(N)
	int s = 0 ; // O(1)
	int r = 0 ; // O(1)
	
	for (int i = 0 ; i < N ; i++){ // O(N)
	    s = max(s+v[i],v[i]) ; // O(1)
	    r = max(r,s) ; // O(1)
    }
    
    return r ; // O(1)
}

int main()
{
    
    /* EJECUCIÓN DE LOS PROGRAMAS */
    
    const int DIM = 50 ;
    int v[DIM] ;
    int N ;
   
    cout << "Introduce la longitud del vector: " ;
    cin >> N ;
    cout << endl ;
    
    cout << "Introduce las " << N << " componentes del vector: \n" ;
    for (int i = 0 ; i < N ; i++) cin >> v[i] ;
    cout << endl ;

    cout << "Vector = " ;
    for (int i = 0 ; i < N ; i++) cout << v[i] << " " ;
    cout << endl << endl ;
    cout << "Resultado de la funcion 'queHace1': " << queHace1(v, N) << "\n" ;
    cout << endl ;
    cout << "Resultado de la funcion 'queHace2': " << queHace2(v, N) << "\n" ;
    cout << endl ;
    cout << "Resultado de la funcion 'queHace3': " << queHace3(v, N) << "\n" ;
    cout << endl ;


    /* MEDIDAS DEL TIEMPO DE EJECUCION DE LOS PROGRAMAS */
  
    
    int repeticiones ;
    int tamanyos[DIM] ;
    double tiempos1[DIM] ;
    clock_t t1, t2 ;
    
    cout << "Introduce el numero de repeticiones: " ;
    cin >> repeticiones ;
    cout << endl ;

    for (int i = 0 ; i < N ; i++) tamanyos[i] = i + 1 ;
  
    cout << "Medidas del tiempo de ejecucion del programa 'queHace1': \n" << endl ;

    for (int k = 0 ; k < N ; k++){
        
       t1 = clock() ;
       
       for (int i = 0 ; i < repeticiones ; i++){
           
          queHace1(v, tamanyos[k]) ;
          
       }
       
       t2 = clock() ;
       
       tiempos1[k] = (double(t2 - t1) / CLOCKS_PER_SEC) / repeticiones ;
       
    }
    
    for (int k = 0 ; k < N ; k++)
         cout << "Longitud = " << tamanyos[k] << "; Tiempo = " << tiempos1[k] << endl ;
    
    cout << endl ;
    
    
    
    /*
     
     EJERCICIO 3: Escribe el código en C++ para obtener las medidas del tiempo de ejecución
     de los programas 'queHace2' y "queHace3', y compara los resultados obtenido con los del
     programa 'queHace1'. ¿Cuál de los tres programas es más eficiente? ¿Por qué?
     
     */
	
	double tiempos2[DIM] ;
	double tiempos3[DIM] ;
	
	/* queHace2 */ 
	
    cout << "Medidas del tiempo de ejecucion del programa 'queHace2': \n" << endl ;

    for (int k = 0 ; k < N ; k++){
        
       t1 = clock() ;
       
       for (int i = 0 ; i < repeticiones ; i++){
           
          queHace2(v, tamanyos[k]) ;
          
       }
       
       t2 = clock() ;
       
       tiempos2[k] = (double(t2 - t1) / CLOCKS_PER_SEC) / repeticiones ;
       
    }
    
    for (int k = 0 ; k < N ; k++)
         cout << "Longitud = " << tamanyos[k] << "; Tiempo = " << tiempos2[k] << endl ;
    
    cout << endl ;
    
    /* queHace3 */
    
    cout << "Medidas del tiempo de ejecucion del programa 'queHace3': \n" << endl ;

    for (int k = 0 ; k < N ; k++){
        
       t1 = clock() ;
       
       for (int i = 0 ; i < repeticiones ; i++){
           
          queHace3(v, tamanyos[k]) ;
          
       }
       
       t2 = clock() ;
       
       tiempos3[k] = (double(t2 - t1) / CLOCKS_PER_SEC) / repeticiones ;
       
    }
    
    for (int k = 0 ; k < N ; k++)
         cout << "Longitud = " << tamanyos[k] << "; Tiempo = " << tiempos3[k] << endl ;
    
    cout << endl ;

	/* Vamos a utilizar el vector con mayor longitud para comparar el tiempo de ejecucion de cada programa */
	
	double n1 = tiempos1[N-1] ;
	double n2 = tiempos2[N-1] ;
	double n3 = tiempos3[N-1] ;
	
	if ( min( min(n1, n2), n3) == n1 ) {
		cout << "El tiempo de ejecucion del programa 'queHace1' es el menor, por lo tanto es el programa mas eficiente" << endl ;	
	}
	
	else if ( min( min(n1, n2), n3) == n2 ) {
		cout << "El tiempo de ejecucion del programa 'queHace2' es el menor, por lo tanto es el programa mas eficiente" << endl ;	
	}
	
	else {
		cout << "El tiempo de ejecucion del programa 'queHace3' es el menor, por lo tanto es el programa mas eficiente" << endl ;
	}
}
